teps/0135-coscheduling-pipelinerun-pods.md

```diff
@@ -0,0 +1,412 @@
+
+### Existing Workarounds
+
+Users can rewrite their PipelineRuns to use a single PVC-backed workspace and multiple sub-paths, with the affinity assistant enabled. See [this comment](https://github.com/tektoncd/pipeline/issues/3480#issuecomment-720442858) for more detail.
+
+### Requirements
+
+- Must be possible for the cluster operator to configure whether multiple PipelineRuns
+can run concurrently on the same node.
+- If it's not possible for all of a PipelineRun's pods to be scheduled to a node, the PipelineRun must still run to completion without operator intervention. See [Unschedulable Nodes](#unschedulable-nodes) for more detail.
+
+## Proposal
+
+PipelineRun pods can be scheduled to the same node using a mechanism similar to the existing affinity assistant: creating a placeholder pod and setting [inter-pod affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) for TaskRun pods to the placeholder pod. While the existing affinity assistant creates one placeholder pod per workspace, the proposed solution will create one placeholder pod per PipelineRun.
+
+## Design Details
+
+### Configuration
+
+Currently, the affinity assistant is enabled by default, and disabled by setting the boolean feature flag "disable-affinity-assistant" to true.
+
+This TEP proposes creating a new feature flag, "coscheduling". The options will be:
+- "disabled": The default value.
+- "coschedule-workspaces": The existing affinity assistant behavior. This option runs all of a PipelineRun's pods sharing the same workspace on the same node.
+- "coschedule-pipelineruns": This option runs all of a PipelineRun's pods on the same node.
+- "isolate-pipelineruns": This option runs all of a PipelineRun's pods on the same node, and only allows one PipelineRun to run on a node at a time.
+
+If the existing affinity assistant is enabled, the "coscheduling" flag must be set to "disabled". We will deprecate the "disable-affinity-assistant" flag in favor of the "coscheduling" flag, as discussed in ["Upgrade and Migration Strategy"](#upgrade-and-migration-strategy).
+
+### Placeholder pod + inter-pod affinity
+
+The proposed implementation is very similar to the existing affinity assistant. Using a [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/), the PipelineRun controller will create a placeholder pod per PipelineRun. All TaskRun pods will be modified to have inter-pod affinity for the placeholder pod. If the "isolate-pipelineruns" configuration option is chosen, the placeholder pod will have inter-pod anti-affinity for other placeholder pods. The placeholder pod does nothing, has minimal resource requirements, and is torn down on PipelineRun completion.
+
+#### Example: Coscheduling a PipelineRun
+
+The placeholder pod will have the following labels:
+
+```
+app.kubernetes.io/component: per-pipelinerun-affinity-assistant
```

Comment:
Do we want to still use the "affinity-assistant" term here (as it disappears from the rest for `coscheduling`) ?

Reply:
I think it's probably fine since the placeholder pod is an implementation detail and users aren't interacting with it. it seems easier to just keep the existing labels, but we can definitely change them if that's the consensus.